x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/InferenceUtils.java

```diff
@@ -0,0 +1,124 @@
+        BiConsumer<LogicalPlan, ActionListener<Result>> callback
+    ) {
+        Set<SemanticQuery> semanticQueries = semanticQueries(plan);
+
+        if (semanticQueries.isEmpty()) {
+            callback.accept(plan, listener);
+            return;
+        }
+
+        GroupedActionListener<InferenceAction.Response> actionListener = new GroupedActionListener<>(
+            semanticQueries.size(),
+            new ActionListener<Collection<InferenceAction.Response>>() {
+                @Override
+                public void onResponse(Collection<InferenceAction.Response> ignored) {
+                    try {
+                        callback.accept(plan, listener);
+                    } catch (Exception e) {
+                        onFailure(e);
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    listener.onFailure(e);
+                }
+            }
+        );
+        for (SemanticQuery semanticQuery : semanticQueries) {
+            InferenceAction.Request inferenceRequest = new InferenceAction.Request(
+                TaskType.ANY,
+                semanticQuery.inferenceId(),
+                null,
+                List.of(semanticQuery.queryString()),
+                Map.of(),
+                InputType.SEARCH,
+                InferModelAction.Request.DEFAULT_TIMEOUT_FOR_API,
+                false
+            );
+
+            executeAsyncWithOrigin(
+                client,
+                ML_ORIGIN,
+                InferenceAction.INSTANCE,
+                inferenceRequest,
+                actionListener.delegateFailureAndWrap((next, inferenceResponse) -> {
+                    InferenceResults inferenceResults = SemanticTextInferenceUtils.validateAndConvertInferenceResults(
+                        inferenceResponse.getResults(),
+                        semanticQuery.fieldName()
+                    );
+                    setInferenceResult(plan, semanticQuery.fieldName(), semanticQuery.queryString(), inferenceResults);
+                    next.onResponse(inferenceResponse);
+                })
+            );
+        }
+    }
+
+    private record SemanticQuery(String fieldName, String queryString, String inferenceId) {};
+
+    private static Set<SemanticQuery> semanticQueries(LogicalPlan analyzedPlan) {
+        Set<SemanticQuery> result = new HashSet<>();
+
+        analyzedPlan.forEachExpressionDown(Match.class, match -> {
+            if (match.field().dataType() == DataType.SEMANTIC_TEXT && match.field() instanceof FieldAttribute) {
+                FieldAttribute field = (FieldAttribute) match.field();
+                SemanticTextEsField esField = (SemanticTextEsField) field.field();
+                if (esField.inferenceIds().size() == 1) {
+                    result.add(new SemanticQuery(field.sourceText(), match.query().sourceText(), esField.inferenceIds().iterator().next()));
```

Comment:
> In case there are multiple inferenceIds, do we want to throw an exception here? Similar with what is done in Verifier.

I guess it does not hurt to throw an exception - but at this point in the execution, the `Verifier` already checked and returned an error if we are dealing with multiple inference ids.